File: src/mtglab/auth/passwords.py
"""Argon2id, at the OWASP minimum profile, and nothing else.

The parameters are not a default to tune later — they are the decision in
ADR 5, and the reason is memory rather than security. Argon2 is deliberately
memory-hard, so the widely-copied `m=64MB` setting means **each concurrent
login allocates 64 MB**. On the 1 GB instance `docs/HOSTING.md` §4 sizes, a
handful of simultaneous logins alongside DuckDB and a running simulation is how
the machine gets OOM-killed. `m=19MiB, t=2, p=1` is OWASP's stated minimum, is
still far beyond brute-force economics for a private site with a dozen
accounts, and cannot exhaust the box.

The other thing here is `verify_dummy()`, which exists to be *wasted*. If an
unknown username returns before a hash is computed and a known one returns
after, the response time answers "does this account exist" for anybody with a
stopwatch — and answering that is the same leak ADR 16 forbids the reset
endpoint from giving away through the front door.
"""

from __future__ import annotations

from contextlib import suppress
from functools import cache

from argon2 import PasswordHasher
from argon2.exceptions import (
    InvalidHashError,
    VerificationError,
    VerifyMismatchError,
)

from mtglab import caches

# OWASP minimum profile. `memory_cost` is in KiB: 19456 KiB == 19 MiB.
MEMORY_COST_KIB = 19_456
TIME_COST = 2
PARALLELISM = 1

# A password field long enough to be a denial of service is a password field
# with no upper bound: Argon2 hashes its input at whatever length it arrives,
# and 19 MiB of memory per attempt is already the expensive part. 1024 bytes is
# far past any real passphrase and far short of anything worth worrying about.
MAX_PASSWORD_BYTES = 1024

# The floor is a floor, not advice. Long is what matters; complexity rules push
# people toward `Passw0rd!` and this project has no interest in that argument.
MIN_PASSWORD_LENGTH = 12


class WeakPassword(ValueError):
    """Raised rather than storing a password the caller should have refused."""


@cache
def _hasher() -> PasswordHasher:
    return PasswordHasher(time_cost=TIME_COST, memory_cost=MEMORY_COST_KIB,
                          parallelism=PARALLELISM)


@cache
def _dummy_hash() -> str:
    """A real hash of a value nobody knows, computed once and reused.

    Once, not per call: computing it fresh each time would cost two hashes on
    an unknown user against one on a known user, which is the same timing
    signal pointing the other way.
    """
    return _hasher().hash("mtglab-dummy-password-for-timing-parity")


def check_strength(password: str) -> None:
    """Raise `WeakPassword` if this is not allowed to be stored."""
    if len(password.encode("utf-8")) > MAX_PASSWORD_BYTES:
        raise WeakPassword(
            f"password is longer than {MAX_PASSWORD_BYTES} bytes")
    if len(password) < MIN_PASSWORD_LENGTH:
        raise WeakPassword(
            f"password must be at least {MIN_PASSWORD_LENGTH} characters "
            "-- length beats punctuation, so a short phrase is a fine answer")


def hash_password(password: str) -> str:
    """Hash a password for storage. Refuses one that should not be stored."""
    check_strength(password)
    return _hasher().hash(password)


def verify(stored_hash: str | None, password: str) -> bool:
    """Is this the password behind that hash?

    `None` -- an account that has been created but has no password yet -- costs
    a dummy verification and returns False, so an invited-but-unclaimed account
    is not identifiable by how fast it is refused.

    Every argon2 failure mode collapses to `False`: a mismatch, a corrupt hash,
    a hash produced by some other library. There is no caller that wants to
    tell those apart at the point of a login, and one that returned a different
    answer for "malformed stored hash" would be reporting on the database to
    whoever typed the password.
    """
    if stored_hash is None:
        verify_dummy(password)
        return False
    try:
        return _hasher().verify(stored_hash, password)
    except (VerifyMismatchError, VerificationError, InvalidHashError):
        return False


def verify_dummy(password: str) -> None:
    """Burn a verification against a hash of nothing. For unknown accounts.

    The result is thrown away -- the work is the whole point. It will not
    match, because nobody knows the string it was made from.
    """
    with suppress(VerifyMismatchError, VerificationError, InvalidHashError):
        _hasher().verify(_dummy_hash(), password)


def needs_rehash(stored_hash: str) -> bool:
    """Was this hash made with weaker parameters than the ones above?

    Checked at login, when the plaintext is in hand and re-hashing is free --
    the only moment an old hash can be upgraded without asking anybody to do
    anything.
    """
    try:
        return _hasher().check_needs_rehash(stored_hash)
    except InvalidHashError:
        return False


#: Both of these are `functools.cache`, which counts itself -- so the register
#: only has to be told they exist. `_dummy_hash`'s docstring makes a claim
#: worth checking rather than trusting ("computed once and reused"): a hit rate
#: below ~100% across a run of failed logins would mean the timing defence is
#: paying for two hashes where it budgeted for one.
caches.register_lru("auth.hasher", _hasher,
                    note="the Argon2 hasher, built once per process")
caches.register_lru("auth.dummy-hash", _dummy_hash,
                    note="the throwaway hash an unknown account is verified "
                         "against, so login costs the same either way")

File: go/internal/auth/passwords.go
package auth

import (
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"github.com/alexedwards/argon2id"
	"golang.org/x/crypto/argon2"
)

// The OWASP minimum profile, pinned: memory-hard
// enough to matter, small enough (19 MiB a call) that a handful of concurrent
// logins cannot OOM the 1GB instance. Not a default to tune; the decision is
// ADR 5's.
const (
	MemoryCostKiB = 19_456
	TimeCost      = 2
	Parallelism   = 1
	saltLength    = 16 // the recorded hashes' salt length
	keyLength     = 32 // the recorded hashes' digest length
)

// MaxPasswordBytes bounds the input a hash will be computed over, and
// MinPasswordLength is the floor -- length, not complexity.
const (
	MaxPasswordBytes  = 1024
	MinPasswordLength = 12
)

// ErrWeakPassword is raised rather than storing a password that should have
// been refused.
var ErrWeakPassword = errors.New("weak password")

var params = &argon2id.Params{
	Memory:      MemoryCostKiB,
	Iterations:  TimeCost,
	Parallelism: Parallelism,
	SaltLength:  saltLength,
	KeyLength:   keyLength,
}

// CheckStrength refuses a password that must not be stored.
func CheckStrength(password string) error {
	if len(password) > MaxPasswordBytes {
		return failf("%w: password is longer than %d bytes",
			ErrWeakPassword, MaxPasswordBytes)
	}
	if len([]rune(password)) < MinPasswordLength {
		return failf("%w: password must be at least %d characters "+
			"-- length beats punctuation, so a short phrase is a fine answer",
			ErrWeakPassword, MinPasswordLength)
	}
	return nil
}

// HashPassword hashes for storage, in the recorded PHC string form
// (`$argon2id$v=19$m=19456,t=2,p=1$<salt>$<hash>`), so a hash made today
// is interchangeable with every hash already in the file.
//
// The salt is drawn here and the encoding is `hashWithSalt` below rather than
// the library's own `CreateHash`, and that is deliberate: it makes the salt an
// *argument*, which is what lets `testdata/crypto.json` pin this function's
// output **byte for byte** against the recorded corpus instead of settling
// for a round trip. A hash written here has to verify for the rest of the
// file's life, whatever verifies it next.
func HashPassword(password string) (string, error) {
	if err := CheckStrength(password); err != nil {
		return "", err
	}
	salt := make([]byte, saltLength)
	if _, err := rand.Read(salt); err != nil {
		return "", fmt.Errorf("no entropy for a password salt: %w", err)
	}
	return hashWithSalt(password, salt), nil
}

// hashWithSalt is the encoder: Argon2id at the profile above, written as the
// recorded PHC string.
//
// Every detail here is the format's decision, copied rather than
// chosen. The version field is `v=19` (0x13), the parameter order is `m,t,p`,
// and the salt and digest are **unpadded** standard base64 -- not the URL
// alphabet, which is the one mistake that would produce a string a strict
// PHC parser accepts for some salts and rejects for others, so the corpus
// carries a salt with both `+` and `/` in it.
func hashWithSalt(password string, salt []byte) string {
	sum := argon2.IDKey([]byte(password), salt, TimeCost, MemoryCostKiB,
		Parallelism, keyLength)
	return fmt.Sprintf("$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
		argon2.Version, MemoryCostKiB, TimeCost, Parallelism,
		base64.RawStdEncoding.EncodeToString(salt),
		base64.RawStdEncoding.EncodeToString(sum))
}

// Verify asks whether password is the one behind storedHash. A nil hash --
// an invited account with no password yet -- costs a dummy verification and
// answers false, so an unclaimed account is not identifiable by how fast it
// is refused. Every failure mode collapses to false: a
// mismatch, a corrupt hash, a hash from some other library.
func Verify(storedHash *string, password string) bool {
	if storedHash == nil {
		VerifyDummy(password)
		return false
	}
	ok, err := argon2id.ComparePasswordAndHash(password, *storedHash)
	return err == nil && ok
}

// dummyHash is computed once per process:
// computing it per call would cost two hashes on an unknown account against
// one on a known one, the same timing signal pointing the other way.
var dummyHash = func() string {
	h, err := argon2id.CreateHash("mtglab-dummy-password-for-timing-parity", params)
	if err != nil {
		panic(err)
	}
	return h
}()

// VerifyDummy burns a verification against a hash of nothing, for unknown
// accounts. The result is thrown away; the work is the point.
func VerifyDummy(password string) {
	_, _ = argon2id.ComparePasswordAndHash(password, dummyHash)
}

// NeedsRehash reports whether a stored hash was made with weaker parameters
// than the ones above -- checked at login, when the plaintext is in hand.
func NeedsRehash(storedHash string) bool {
	p, _, _, err := argon2id.DecodeHash(storedHash)
	if err != nil {
		return false
	}
	return p.Memory < MemoryCostKiB || p.Iterations < TimeCost ||
		p.Parallelism < Parallelism || !strings.HasPrefix(storedHash, "$argon2id$")
}
